### PR TITLE
Modify esmeralda

### PR DESCRIPTION
**Pull request by mmkekic**
This PR adds no-filter KDST/MC/Run tables and removes corresponding duplicated variable from PAOLINA/Summary table, as discussed in #671

### CODE REVIEW

**Comment by jahernando** [comment]
This PR modifies esmeralda to output run_info and kdst without filters; reduces the common columns in summary and kdst, and provides the user a tool to extend the summary with variables from the kdst. It modifies the test accordingly.

The code is compact and clear, a little cryptic for those not so familiar with the capabilities of pandas. It will help to future code maintainers, to add some comments in plain english to translate what it is done in the code, sometimes, in one line, for example in the *load_event_summary* function.

If not other revisions of variables to be added or removed are needed in Esmeralda. I approve this PR.

**Review by jahernando — APPROVED** [review]
This PR modifies esmeralda to output run_info and kdst without filters; reduces the common columns in summary and kdst, and provides the user a tool to extend the summary with variables from the kdst. It modifies the test accordingly.

The code is compact and clear, a little cryptic for those not so familiar with the capabilities of pandas. It will help to future code maintainers, to add some comments in plain english to translate what it is done in the code, sometimes, in one line, for example in the load_event_summary function.

If not other revisions of variables to be added or removed are needed in Esmeralda. I approve this PR.

File: invisible_cities/reco/dst_functions.py [hunk 1]
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
...
         return data
+
+
+def load_event_summary(filename : str) -> pd.DataFrame :
+    summary = load_dst(filename, 'PAOLINA', 'Summary')
+    kdst    = load_dst(filename, 'DST'    , 'Events' )
+    print(kdst.columns)
+    if(len(kdst.s1_peak.unique()) != 1):
+        warnings.warn("Number of recorded S1 energies differs from 1 in event {}.Choosing first S1".format(event_number), UserWarning)
+    kdst_to_merge = kdst[['event', 'S1e', 'S1t', 'nS2', 'S2e', 'S2q']].groupby('event').agg({'S1e':lambda x:x.values[0],
```
Comment (gonzaponte) [inline invisible_cities/reco/dst_functions.py]: `kdst.loc[:, columns]` is better (returns a view, not a copy)

File: invisible_cities/cities/esmeralda.py [hunk 2]
```diff
@@ -277,27 +276,16 @@ def make_event_summary(event_number  : int              ,
...
     """
-    es = pd.DataFrame(columns=['event', 'time', 'S1e', 'S1t',
-                               'nS2', 'ntrks', 'nhits', 'S2e0', 'S2ec',
-                               'S2q0', 'S2qc', 'x_avg', 'y_avg', 'z_avg',
-                               'r_avg', 'x_min', 'y_min', 'z_min', 'r_min',
-                               'x_max', 'y_max', 'z_max', 'r_max'])
+    es = pd.DataFrame(columns=['event', 'time', 'S2ec', 'S2qc', 'ntrks',
+                               'nhits', 'x_avg', 'y_avg', 'z_avg','r_avg',
+                               'x_min', 'y_min', 'z_min', 'r_min','x_max',
+                               'y_max', 'z_max', 'r_max'])
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: minor missalignment

File: invisible_cities/reco/dst_functions_test.py [hunk 3]
```diff
@@ -86,3 +91,13 @@ def test_dst_event_id_selection_2():
     df_real_filt = dst_event_id_selection(df_data, [1, 2, 6, 10])
 
     assert np.all(df_real_filt.reset_index(drop=True) == df_filt_data.reset_index(drop=True))
+
+
+def test_load_event_summary_equal_to_old_summary(ICDATADIR):
+    old_esmeralda   =  os.path.join(ICDATADIR, "exact_Kr_tracks_with_MC.h5")
+    kdst_esmeralda =  os.path.join(ICDATADIR, "exact_Kr_tracks_with_MC_KDST.h5")
```
Comment (gonzaponte) [inline invisible_cities/reco/dst_functions_test.py]: minor missalignment

File: invisible_cities/reco/dst_functions.py [hunk 4]
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
     else:
         print(r'DST does not have an "event" field. Data returned is unfiltered.')
         return data
+
+
+def load_event_summary(filename : str) -> pd.DataFrame :
+    summary = load_dst(filename, 'PAOLINA', 'Summary')
+    kdst    = load_dst(filename, 'DST'    , 'Events' )
+    print(kdst.columns)
```
Comment (gonzaponte) [inline invisible_cities/reco/dst_functions.py]: forgotten print?

File: invisible_cities/reco/dst_functions.py [hunk 5]
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
...
+    if(len(kdst.s1_peak.unique()) != 1):
+        warnings.warn("Number of recorded S1 energies differs from 1 in event {}.Choosing first S1".format(event_number), UserWarning)
```
Comment (gonzaponte) [inline invisible_cities/reco/dst_functions.py]: Do you have a protection agains nS1 = 0?

File: invisible_cities/reco/dst_functions.py [hunk 6]
```diff
@@ -72,3 +73,21 @@ def dst_event_id_selection(data, event_ids):
...
+    kdst_to_merge = kdst[['event', 'S1e', 'S1t', 'nS2', 'S2e', 'S2q']].groupby('event').agg({'S1e':lambda x:x.values[0],
+                                                                                             'S1t':lambda x:x.values[0],
+                                                                                             'nS2':lambda x:x.values[0],
+                                                                                             'S2e' : np.sum,
+                                                                                             'S2q' : np.sum}).reset_index()
+    kdst_to_merge.columns = ['event', 'S1e', 'S1t', 'nS2', 'S2e0', 'S2q0']
+    print(kdst_to_merge.columns)
```
Comment (gonzaponte) [inline invisible_cities/reco/dst_functions.py]: another forgotten print?

File: invisible_cities/reco/monitor_functions_test.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
 from hypothesis              import given
 from hypothesis              import settings
 from hypothesis.extra.pandas import columns, data_frames
```
Comment (gonzaponte) [inline invisible_cities/reco/monitor_functions_test.py]: it's not part of the PR, but please split this double import into two lines :)

File: invisible_cities/reco/monitor_functions_test.py [hunk 8]
```diff
@@ -416,7 +417,7 @@ def test_kdst_bins():
                   'Xrms', 'Yrms']
 
 
-@given(data_frames(columns=columns(kdst_variables, elements=floats(allow_nan=False))))
+@given(data_frames(index=range_indexes(min_size=1, max_size=10), columns=columns(kdst_variables, elements=floats(allow_nan=False))))
```
Comment (gonzaponte) [inline invisible_cities/reco/monitor_functions_test.py]: I think it would be good to define somewhere `kdst_dataframes = data_frames(...)` (or similar) for readability and to avoid repetition.

File: invisible_cities/cities/esmeralda.py [hunk 9]
```diff
@@ -418,19 +417,14 @@ def esmeralda(files_in, file_out, compression, event_range, print_mod, run_numbe
...
     paolina_params_    = {value : paolina_params .get(value) for value in ['vox_size' , 'strict_vox_size', 'energy_threshold', 'min_voxels', 'blob_radius']}
 
-    threshold_and_correct_hits_NN      = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_NN'    ], norm_strat = norm_strat),
+    threshold_and_correct_hits_NN      = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_NN'    ]),
                                                 args = 'hits',
                                                 out  = 'NN_hits')
 
-    threshold_and_correct_hits_paolina = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_paolina'], norm_strat = norm_strat),
+    threshold_and_correct_hits_paolina = fl.map(hits_threshold_and_corrector(**cor_hits_params_, threshold_charge = cor_hits_params['threshold_charge_paolina']),
                                                 args = 'hits',
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: Not a big deal, but the convention is to add spaces around equal signs in keyword arguments when each keyword is in a separate line, i.e.
`im_a_function(a, b, c=something, d=something_else)`
or
```
im_a_function(a, b,
              c = something,
              d = something_else)
```

Also, usually `**kwargs` go after hardcoded arguments. Like
`im_a_function(a, b, *c, d=something, **e)`

This happens in more places in the code, but I didn't mark them all. As I say is not a big deal, not worth the effort to look for them, but it is good to keep it in mind for future modifications to the code.

File: invisible_cities/cities/esmeralda.py [hunk 10]
```diff
@@ -91,8 +91,6 @@ def threshold_and_correct_hits(hitc : evm.HitCollection) -> evm.HitCollection:
         t = hitc.time
         thr_hits = hif.threshold_hits(hitc.hits, threshold_charge     )
         mrg_hits = hif.merge_NN_hits ( thr_hits, same_peak = same_peak)
-        if len(mrg_hits) == 0:
-            return None
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py:96]: is this safe?

File: invisible_cities/cities/esmeralda.py [hunk 11]
```diff
@@ -163,20 +145,16 @@ class HitEnergy(AutoNameEnumBase):
...
-        tracks     = sorted(tracks, key = get_track_energy, reverse = True)
-
-        track_hits = []
+        #track_hits is a new Hitcollection object that contains hits belonging to tracks, and hits that couldnt be corrected
+        track_hitc = evm.HitCollection(hitc.event, hitc.time)
+        out_of_map = any(np.isnan([h.Ec for h in hitc.hits]))
+        if out_of_map:
+            #add nan hits to track_hits, the track_id will be -1
+            track_hitc.hits.extend([h for h in hitc.hits if np.isnan   (getattr(h, energy_type.value))])
+            hits_without_nan       = [h for h in hitc.hits if np.isfinite(getattr(h, energy_type.value))]
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: alignment

File: invisible_cities/cities/esmeralda.py [hunk 12]
```diff
@@ -186,63 +164,78 @@ def get_track_energy(track):
...
+
+                min_x = min([h.X for v in t.nodes() for h in v.hits])
+                max_x = max([h.X for v in t.nodes() for h in v.hits])
+                min_y = min([h.Y for v in t.nodes() for h in v.hits])
+                max_y = max([h.Y for v in t.nodes() for h in v.hits])
+                min_z = min([h.Z for v in t.nodes() for h in v.hits])
+                max_z = max([h.Z for v in t.nodes() for h in v.hits])
+                max_r = max([np.sqrt(h.X*h.X + h.Y*h.Y) for v in t.nodes() for h in v.hits])
+
+                pos = [h.pos for v in t.nodes() for h in v.hits]
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: I've seen this kind of thing in several places... This is a double loop repeated 7 times. I haven't time-profiled it, but my feeling is that it is very inefficient. I think it is worth to make a function that makes this loop only once and retrieves all the variables that may be needed. Opinions?

File: invisible_cities/cities/esmeralda_test.py [hunk 13]
```diff
@@ -128,6 +128,10 @@ def test_esmeralda_with_out_of_map_hits(KrMC_hdst_filename_toy, correction_map_M
     assert set(df_hits_low_th     .event.unique()) ==  set(events_pass_low_th     )
     assert set(df_hits_paolina.event.unique()) ==  set(events_pass_paolina)
 
+    summary_table       =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Summary')
+    #assert event with nan energy labeled in summary_table
+    events_energy = df_hits_paolina.groupby('event')['Ec'].apply(pd.Series.sum, skipna=False)
+    np.testing.assert_array_equal( summary_table['out_of_map'], np.isnan(events_energy.values))
     #assert event number in EventInfo and MC/Extents iqual to nevt_req
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: `.Ec` instead of `['Ec']`?
`.out_of_map` instead of `['out_of_map']`?

File: invisible_cities/cities/esmeralda_test.py [hunk 14]
```diff
@@ -284,3 +284,39 @@ def test_esmeralda_blob_overlap_bug(data_hdst, esmeralda_tracks, correction_map_
...
+
+    cnt = esmeralda(**conf)
+
+    tables = ["MC/extents", "MC/hits", "MC/particles",
+              "PAOLINA/Events", "PAOLINA/Summary", "PAOLINA/Tracks", "RECO/Events",
+              "Run/events", "Run/runInfo", "DST/Events",
+              "Filters/low_th_select", "Filters/high_th_select", "Filters/topology_select"]
+
+    with tb.open_file(true_out)  as true_output_file:
+        with tb.open_file(file_out) as      output_file:
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: alignment

File: invisible_cities/cities/esmeralda.py [hunk 15]
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
...
+            voxels     = plf.voxelize_hits(hitc.hits, vox_size, strict_vox_size, evm.HitEnergy.Ec)
+            mod_voxels = plf.drop_end_point_voxels(voxels, energy_threshold, min_voxels)
+            tracks     = plf.make_track_graphs(mod_voxels)
+
+            vox_size_x = voxels[0].size[0]
+            vox_size_y = voxels[0].size[1]
+            vox_size_z = voxels[0].size[2]
+            del(voxels)
+            def get_track_energy(track):
+                return sum([vox.Ehits for vox in track.nodes()])
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: this function here looks weird.. Isn't there a module for track functions?

File: invisible_cities/cities/esmeralda.py [hunk 16]
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
...
+            tracks     = sorted(tracks, key = get_track_energy, reverse = True)
+
+            track_hits = []
+            for c, t in enumerate(tracks, 0):
+                tID = c
+                energy = get_track_energy(t)
+                length = plf.length(t)
+                numb_of_hits = len([h for vox in t.nodes() for h in vox.hits])
+                numb_of_voxels = len(t.nodes())
+                numb_of_tracks = len(tracks   )
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: alignment

File: invisible_cities/cities/esmeralda.py [hunk 17]
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
...
+                ave_pos = np.average(pos, weights=e, axis=0)
+
+                extr1, extr2 = plf.find_extrema(t)
+                extr1_pos = extr1.XYZ
+                extr2_pos = extr2.XYZ
+
+                blob_pos1, blob_pos2 = plf.blob_centres(t, blob_radius)
+
+                e_blob1, e_blob2, hits_blob1, hits_blob2 = plf.blob_energies_and_hits(t, blob_radius)
+                overlap = float(sum([h.Ec for h in set(hits_blob1).intersection(hits_blob2)]))
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: shouldn't it be `set(hits_blob1).intersection(set(hits_blob2))`?

File: invisible_cities/cities/esmeralda.py [hunk 18]
```diff
@@ -187,73 +157,86 @@ def get_track_energy(track):
...
+                list_of_vars = [hitc.event, tID, energy, length, numb_of_voxels,
+                                numb_of_hits, numb_of_tracks,
+                                min_x, min_y, min_z, max_x, max_y, max_z, max_r,
+                                ave_pos[0], ave_pos[1], ave_pos[2],
+                                extr1_pos[0], extr1_pos[1], extr1_pos[2],
+                                extr2_pos[0], extr2_pos[1], extr2_pos[2],
+                                blob_pos1[0], blob_pos1[1], blob_pos1[2],
+                                blob_pos2[0], blob_pos2[1], blob_pos2[2],
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: `*extr1_pos, *extr2_pos, *blob_pos1, *blob_pos2` for readability?

File: invisible_cities/cities/esmeralda_test.py [hunk 19]
```diff
@@ -98,29 +108,41 @@ def test_esmeralda_with_out_of_map_hits(KrMC_hdst_filename_toy, correction_map_M
...
     df_hits_paolina     =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Events')
 
-    assert set(df_hits_NN     .event.unique()) ==  set(events_pass_NN     )
+    assert set(df_hits_low_th     .event.unique()) ==  set(events_pass_low_th     )
     assert set(df_hits_paolina.event.unique()) ==  set(events_pass_paolina)
 
+    summary_table       =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Summary')
+    #assert event with nan energy labeled in summary_table
+    events_energy = df_hits_paolina.groupby('event')['Ec'].apply(pd.Series.sum, skipna=False)
+    np.testing.assert_array_equal( summary_table['out_of_map'], np.isnan(events_energy.values))
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: Use `.column` instead of `[column]`?

File: invisible_cities/cities/esmeralda_test.py [hunk 20]
```diff
@@ -151,11 +171,18 @@ def test_esmeralda_tracks_exact(data_hdst, esmeralda_tracks, correction_map_file
...
+    #some events are not in df_tracks_exact
+    events = df_tracks_exact.event.unique()
     assert(sorted(columns1) == sorted(columns2))
-    assert_dataframes_close (df_tracks[sorted(columns1)], df_tracks_exact[sorted(columns2)])
+    df_tracks_cut  = df_tracks[df_tracks.event.isin(events)]
+    assert_dataframes_close (df_tracks_cut[sorted(columns1)], df_tracks_exact[sorted(columns2)])
+    #make sure out_of_map is true for events not in df_tracks_exact
+    diff_events = list(set(df_tracks.event.unique()).difference(events))
+    df_summary = dio.load_dst(PATH_OUT, 'PAOLINA', 'Summary')
+    assert all(df_summary[df_summary.event.isin(diff_events)]['out_of_map'])
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: use `.loc`

**Review by gonzaponte — COMMENTED** [review]
`test_esmeralda_exact_result_old` can be removed now or in a future PR. There is a commit in the history that proves the new esmeralda yields the same output and a new test to replace it.

File: invisible_cities/cities/esmeralda.py (hunk 12 above)
Comment (jacg) [reply invisible_cities/cities/esmeralda.py]: > haven't time-profiled it, but my feeling is that it is very inefficient.

The most important rule of optimization: don't optimize anything until you've proven that it matters. That means that you must either prove (by timing) that 

a) this is a bottleneck
b) the alternative implementation is actually faster that the original.

If it's not a bottleneck, but the new version is faster and not significantly uglier, then go ahead. If the new version is uglier and more difficult to understand, then don't use it **unless profiling proves that the change makes a significant improvement in perfomance**.

However, the square brackets inside the `max` and `min` ensure that these will run in linear rather than constant space, which is just a pointless waste ... so get rid of the `[]`s. That is

    fn([ for ... in ... ])

should *always* be written as

    fn( for ... in ... )

or

   fn(( for ... in ... ))

if you like pointless characters (or there are other arguments besides the listcomp / genexpr).

**Comment by gonzaponte** [comment]
Yes, I agree, but doesn't it grind your gears to see those loops repeated
so many times?

File: invisible_cities/cities/esmeralda.py (hunk 12 above)
Comment (jacg) [reply invisible_cities/cities/esmeralda.py]: > Yes, I agree, but doesn't it grind your gears to see those loops repeated
so many times?

It does, it does. A smart enough compiler ...

File: invisible_cities/reco/dst_functions.py (hunk 5 above)
Comment (mmkekic) [reply invisible_cities/reco/dst_functions.py]: No because it is always there, will give NaNs, the same as in input kdst table, guess its fine

File: invisible_cities/cities/esmeralda.py (hunk 10 above)
Comment (mmkekic) [reply invisible_cities/cities/esmeralda.py:96]: yeah, the filter is checking if the len of hits is 0, the test  test_esmeralda_filters_events shows the events are properly filtered out

File: invisible_cities/cities/esmeralda.py (hunk 12 above)
Comment (mmkekic) [reply invisible_cities/cities/esmeralda.py]: I agree! However I realized that we are already constructing pos list to use it in  np.average so I just rearranged the order to first construct pos list and then look for appropriate min/max

File: invisible_cities/cities/esmeralda.py [hunk 21]
```diff
@@ -142,12 +122,6 @@ def track_blob_info_creator_extractor(vox_size         : [float, float, float],
     ----------
     vox_size         : [float, float, float]
         (maximum) size of voxels for track reconstruction
-    energy_type      : HitEnergy
-        class HitEnergy(AutoNameEnumBase):
-            E        = auto()
-            Ec       = auto()
-        energy attribute to use for voxelization/ tracks
-
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py:150]: Is `HitEnergy` still used somewhere else? If not remove it.

File: invisible_cities/cities/esmeralda.py [hunk 22]
```diff
@@ -164,96 +138,102 @@ class HitEnergy(AutoNameEnumBase):
...
+                numb_of_voxels = len(t.nodes())
+                numb_of_tracks = len(tracks   )
+
+                pos   = [h.pos for v in t.nodes() for h in v.hits]
+                min_x = min(pos, key = lambda x : x[0])[0]
+                max_x = max(pos, key = lambda x : x[0])[0]
+                min_y = min(pos, key = lambda x : x[1])[1]
+                max_y = max(pos, key = lambda x : x[1])[1]
+                min_z = min(pos, key = lambda x : x[2])[2]
+                max_z = max(pos, key = lambda x : x[2])[2]
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: mmm, I was thinking about something in the lines of
```
pos = bla
x, y, z = map(np.array, zip(*pos))
r = np.sqrt(x**2 + y**2)
```
and then mins and maxes are trivial.

File: invisible_cities/cities/esmeralda.py [hunk 23]
```diff
@@ -164,96 +138,102 @@ class HitEnergy(AutoNameEnumBase):
...
+                                *ave_pos, ave_r, *extr1_pos,
+                                *extr2_pos, *blob_pos1, *blob_pos2,
+                                e_blob1, e_blob2, overlap,
+                                vox_size_x, vox_size_y, vox_size_z]
+
+                df.loc[c] = list_of_vars
+                try:
+                    types_dict
+                except NameError:
+                    types_dict = dict(zip(df.columns, [type(x) for x in list_of_vars]))
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py:210]: it bothers me a little bit that this thing that is actually constant is being defined here in the middle of everything. But I am not sure the alternative is better. I would try to define the df columns in the top of the file together with this type mapping. BTW, why is it needed?

File: invisible_cities/cities/esmeralda.py [hunk 24]
```diff
@@ -277,52 +257,44 @@ def make_event_summary(event_number  : int              ,
...
-    x_avg = sum([h.X*h.Ec for h in paolina_hits.hits])
-    y_avg = sum([h.Y*h.Ec for h in paolina_hits.hits])
-    z_avg = sum([h.Z*h.Ec for h in paolina_hits.hits])
-    r_avg = sum([(h.X**2 + h.Y**2)**0.5*h.Ec for h in paolina_hits.hits])
+    S2ec  = max(-1, sum(topology_info.energy))
+    #once charge correction gets implemented we probably want to add it to tracks dataframe too
+    try:
+        S2qc  = max(-1, sum(topology_info.charge))
+    except AttributeError:
+        S2qc  = -1
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: `S2qc = max(-1, sum(topology_info.charge)) if hasattr(topology_info, "charge") else -1`?

File: invisible_cities/cities/esmeralda.py [hunk 25]
```diff
@@ -277,52 +257,44 @@ def make_event_summary(event_number  : int              ,
...
+    x_avg = sum(topology_info.x_ave * topology_info.energy)
+    y_avg = sum(topology_info.y_ave * topology_info.energy)
+    z_avg = sum(topology_info.z_ave * topology_info.energy)
+    r_avg = sum(topology_info.r_ave * topology_info.energy)
     if(S2ec > 0):
         x_avg /= S2ec
         y_avg /= S2ec
         z_avg /= S2ec
         r_avg /= S2ec
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda.py]: isn't `np.average` better?

File: invisible_cities/cities/esmeralda_test.py [hunk 26]
```diff
@@ -60,34 +62,42 @@ def test_esmeralda_filters_events(KrMC_hdst_filename_toy, correction_map_MC_file
...
 
+    #assert event number in EventInfo and MC/Extents iqual to nevt_req
+    with tb.open_file(PATH_OUT)  as h5out:
+        event_info = get_event_info(h5out)
+        assert length_of(event_info) == nevt_req
+
+    with tb.open_file(PATH_OUT)  as h5out:
+        MC_num_evs = h5out.root.MC.extents[:]['evt_number']
+        assert len(MC_num_evs) == nevt_req
+
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: why open it twice?

File: invisible_cities/cities/esmeralda_test.py [hunk 27]
```diff
@@ -166,31 +184,56 @@ def test_esmeralda_exact_result(ICDATADIR, KrMC_hdst_filename, correction_map_MC
...
+    #PAOLINA/Tracks table has additional r_min and r_ave columns, so the tables are not equal
+    tracks_out   = dio.load_dst(file_out, 'PAOLINA', 'Tracks')
+    tracks_true  = dio.load_dst(true_out, 'PAOLINA', 'Tracks')
+    assert_dataframes_close(tracks_out[tracks_true.columns], tracks_true)
+
+    #Finally lets confirm Esmeralda contains KDST, RUN and MC table from Penthesilea file
+    tables_in = ( "MC/extents"  , "MC/hits"       , "MC/particles"  , "MC/generators",
+                  "DST/Events"  , "Run/events"    , "Run/runInfo"                    )
+    with tb.open_file(file_in)  as true_output_file:
+        with tb.open_file(file_out) as      output_file:
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: minor missalignment

File: invisible_cities/cities/esmeralda_test.py [hunk 28]
```diff
@@ -200,7 +243,71 @@ def test_esmeralda_empty_input_file(config_tmpdir, ICDATADIR):
...
+                         apply_temp               = False                 ),
+                     paolina_params               = dict(
+                         vox_size                 = [15 * units.mm] * 3    ,
+                         strict_vox_size          = False                  ,
+                         energy_threshold         = 0 * units.keV          ,
+                         min_voxels               = 2                      ,
+                         blob_radius              = 21 * units.mm)        ))
+    cnt = esmeralda(**conf)
+
+    df_tracks           =  dio.load_dst(PATH_OUT, 'PAOLINA', 'Tracks')
```
Comment (gonzaponte) [inline invisible_cities/cities/esmeralda_test.py]: extra spaces that align with nothing

**Review by gonzaponte — COMMENTED** [review]
minor things

File: invisible_cities/cities/esmeralda.py (hunk 23 above)
Comment (gonzaponte) [reply invisible_cities/cities/esmeralda.py:210]: @mmkekic digit: "if i dont change datatype it will be  floats and we dont want that"

**Review by gonzaponte — APPROVED** [review]
Awesome job by @mmkekic to include the new summary structure into esmeralda. Approved!

**Comment by gonzaponte** [comment]
Apparently this needs further work that is a bit orthogonal to this PR. 
#674 and #675 are continuations of this PR.
